rfcs/20220713-sig-open-xla.md

```diff
@@ -0,0 +1,82 @@
+# RFC: Establish SIG OpenXLA
+
+| Status        | Proposed     |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [419](https://github.com/tensorflow/community/pull/419)|
+| **Author(s)** | Thea Lamkin (thealamkin@google.com), Mehdi Amini (aminim@google.com) |
+| **Sponsor**   | Thea Lamkin (thealamkin@google.com)            |
+| **Updated**   | 2022-07-13                                           |
+
+## Objective
+
+OpenXLA will be a community-driven and modular open source compiler. It will enable efficient lowering, optimization and deployment of ML models from most major frameworks to any hardware backend notably CPUs, GPUs, and ML ASICs. This work will be done collaboratively with major ML frameworks and hardware vendors.
+
+SIG OpenXLA will focus on creating the OpenXLA project, including the extraction of XLA from TensorFlow into a standalone project. SIG discussions will facilitate coordination around roadmap, design evolution, and new workflows to be created in OpenXLA. 
```

Comment:
Do you have already an idea of what TF folders will be involved in this process?

Reply:
Also I hope that we are not going to just mirror folders in TF as with the [MHLO repo](https://github.com/tensorflow/mlir-hlo/) and we could clearly isolate the components.

Reply:
The plan on the short-term is to "vendor" OpenXLA inside TensorFlow/third_party, like MLIR was before. Isolating components will require to design stable APIs, release processes, and upgrade processes: this may happen but it'll take time.

MHLO isn't the same setup: it is inside TensorFlow primarily and the standalone repo is a read-only mirror (basically the opposite of venturing).

Reply:
> Isolating components would require to design stable APIs, release processes, and upgrade processes: this may happen in the future but it'll take time and it isn't obvious to do at the C++ level.

I think that this is the most important part. Just with a monolithic approach in `third_party` we are not going to solve the [build invalidation](https://discuss.tensorflow.org/t/llvm-updates-and-bazel-cache/2060) (and some TF breakages) that we experience every day.

> The folders involved are:
....

I've recently contributed to TF2XLA with [many frictions](https://github.com/tensorflow/build/issues/122) between OSS and the internal infra. 
As this folder is not included in your list are these contributions still going to be done in the TF main repo?

Reply:
> I think that this is the most important part. Just with a monolithic approach in third_party we are not going to solve the [build invalidation](https://discuss.tensorflow.org/t/llvm-updates-and-bazel-cache/2060) (and some TF breakages) that we experience every day.

Yes absolutely: this is just an entire different track of work with a different motivation than what motivates OpenXLA right now.
Also, on the topic of build invalidation, LLVM/MLIR will continue to be used in TensorFlow independently of XLA and this won't change. The build invalidation problem will remain an issue for this. I'm not sure what we can do about it though?

> I've recently contributed to TF2XLA with https://github.com/tensorflow/build/issues/122 between OSS and the internal infra.

Ouch... these kind of difference between Bazel and the internal Google checks seems really annoying, we should be able to align this though?

> As this folder is not included in your list are these contributions still going to be done in the TF main repo?

OpenXLA won't have any dependency on TensorFlow, so the TF/XLA bridge will naturally continue to be part of TensorFlow moving forward.
(regardless of where the code go: the kind of problem you refer to will exist and we should address them!)

Reply:
> Also, on the topic of build invalidation, LLVM/MLIR will continue to be used in TensorFlow independently of XLA and this won't change. The build invalidation problem will remain an issue for this. I'm not sure what we can do about it though?

This really depend  on your vision about the productization roadmap.
If TF master/nightly will rely on OpenXLA "rolling sha" commits that will rely on LLVM "rolling sha" commits and so we are really not relying on releases, API versioning, etc.. I think it will be really a weak modularization and not so much that could improve the current status quo.

Some positive side effects could be retreived by disentangling  the targets deps graph: 
https://github.com/tensorflow/community/pull/238

But I think that the main impact is still realated to the OpenXLA own roadmap/vision.

Reply:
> The plan on the short-term is to "vendor" OpenXLA inside TensorFlow/third_party, like MLIR was before it moved to LLVM. Isolating components would require to design stable APIs, release processes, and upgrade processes: this may happen in the future but it'll take time and it isn't obvious to do at the C++ level.
> 
> MHLO isn't the same setup: it is inside TensorFlow primarily and the standalone repo is a read-only mirror (basically the opposite of vendoring).
> 
> The folders involved are:
> 
> * tensorflow/compiler/xla -> will be the new OpenXLA repository root
> * tensorflow/compiler/mlir/hlo -> will move into OpenXLA prior to the split
> * A new "support" repo that will contain the platform abstractions and utilities (tensorflow/core/lib/ and tensorflow/core/platform/ for example, but also some of the profiler runtime).

Hi, I have a question about the plan for the existing XLA compiler (based on HLO IR currently) and MHLO (based on MLIR)?

I see you mentioned that MHLO will also be moved into OpenXLA. What’s the relationship between the XLA compiler and MHLO (based on MLIR) in the future? Will the XLA compiler be re-implemented based on MHLO?

Reply:
> Hi, I have a question about the plan for the existing XLA compiler (based on HLO IR currently) and MHLO (based on MLIR)?
> 
> I see you mentioned that MHLO will also be moved into OpenXLA. What’s the relationship between the XLA compiler and MHLO (based on MLIR) in the future? Will the XLA compiler be re-implemented based on MHLO?

Mostly yes: HLO isn't gonna go away anytime soon, but for the current targets publicly supported by XLA (CPU/GPU) we're pledging to use MLIR (and MHLO) end-to-end on the long term and be the preferred way to add new high-level optimizations to XLA. We're also planning to continue developing most of the codegen inside MLIR/LLVM itself (Linalg in particular) and use it inside XLA. This offers opportunities to share large part of it with other projects like IREE for example.